plutus-scb/src/Cardano/Protocol/Node.hs

```diff
@@ -0,0 +1,217 @@
+
+-- Making Tx work with the node protocols
+type instance HeaderHash Tx = TxId
+type instance HeaderHash Block = BlockId
+deriving instance StandardHash Tx
+
+deriving instance StandardHash Block
+deriving newtype instance NoUnexpectedThunks TxId
+
+type ClientStreamProjection = GlobalStreamProjection ChainState Evt.ChainEvent
+
+chainState :: Lens' ClientStreamProjection ChainState
+chainState = lens getter setter where
+    getter = streamProjectionState
+    setter prj cs = prj { streamProjectionState = cs }
+
+type NodeClientEffects m = '[ChainEffect, State ChainState, Writer [NodeClientEvent], Writer [ChainEvent], EventLogEffect, State ClientStreamProjection, Log, m]
+
+-- Make handleChain use the EventLog.
+type ChainEffs m = '[State ChainState, Writer [ChainEvent], EventLogEffect, m]
+
+-- Note: This is how I would modify the following handlers
+--       to add event log effects
+handleChain' ::
+    ( Members (ChainEffs m) effs
+    , LastMember m effs
+    , MonadEventStore Evt.ChainEvent m )
+ => Eff (ChainEffect ': effs) ~> Eff effs
+handleChain' = interpret $ \case
+    ProcessBlock -> do -- Note: This is something that only the mock server can use.
+        st <- get
+        let pool  = st ^. txPool
+            slot  = st ^. currentSlot
+            idx   = st ^. index
+            (ValidatedBlock block events rest, idx') =
+                validateBlock slot idx pool
+
+        let st' = st & txPool .~ rest
+                   & chainNewestFirst %~ (block :)
+                   & index .~ idx'
+                   & currentSlot +~ 1 -- This assumes that there is exactly one block per slot. In the real chain there may be more than one block per slot.
+        put st'
+        tell events
+        appendEvents
+          [ Evt.NodeEvent $ Evt.BlockAdded block
+          , Evt.NodeEvent $ Evt.NewSlot (fromIntegral $ st' ^. currentSlot) []
+          ]
+
+        pure block
+    QueueTx tx -> do
+        modify $ over txPool (tx :)
+        appendEvent (Evt.NodeEvent $ Evt.SubmittedTx tx)
+
+handleNodeClient' ::
+    ( Members (NodeClientEffects m) effs
+    , LastMember m effs
+    , MonadEventStore Evt.ChainEvent m
+    )
+ => Eff (NodeClientEffect ': effs) ~> Eff effs
+handleNodeClient' = interpret $ \case
+    PublishTx tx   -> queueTx tx >> tell [TxSubmit (txId tx)]
+    GetClientSlot  -> use $ chainState . currentSlot
+    GetClientIndex -> error "Handled by the ChainIndex effect."\
+    -- Note: I am using this to log the newly received block to the event stream.
+    --       I would do away with the Notification and remove CurrentSlot which is
+    --       not used anywhere.
+    ClientNotify n -> case n of
+        BlockValidated blk -> do
+            appendEvent (Evt.NodeEvent $ Evt.BlockAdded blk)
+            get >>= updateProjection
+
+        CurrentSlot sl -> do   -- this is not really needed.
+            appendEvent (Evt.NodeEvent $ Evt.NewSlot sl [])
+            get >>= updateProjection
```

Comment:
Adding the `EventLogEffect` to the base effect handlers (`NodeClientEffect` and `ChainEffect`)

Reply:
@j-mueller Do you also think it is sensible the way I handle log updates?